### ISSUE TITLE
[FEATURE]: Constraints support between columns

### ISSUE THREAD

**Issue by spreeni** [mostly-ai/mostlyai#532]
### Feature Summary

Allow declaration of constraints within columns or across columns that are adhered to in synthesis.

### Problem and Solution

Currently, I believe it is probabilisticly possible that during data synthesis
- an end_time gets sampled before a start_time
- a country + city combination does not match
- increments are not adhering to a standard (e.g. increments of 10m)
- a number is outside of a sensible range (e.g. negative age)

I have moved to mostly ai from SDV due to the permissive license and promising performance and SDV offers column constraints for this:
https://docs.sdv.dev/sdv/concepts/constraint-augmented-generation-cag/predefined-constraints

I think some of these are less troublesome than others and can generally be avoided by oversampling and only selecting items that adhere to the constraints. However, it would be nice to offer this as a convenience functionality in the mostly SDK already - and maybe there even is a more elegant way to let these flow into the modeling already via the loss function or similar approaches.

Thanks for all the beautiful work and for offering under the current license!

**Comment by radurogojanumai** [comment]
Hi @spreeni , thanks for trying out our SDK and for adding your request.

Indeed, even if training performed well, there are non-zero probabilities of getting a combination outside of some constraints that may be apparent in the original data. 
As you mentioned, oversampling and rejecting those that do not follow a certain rule is one possible way to tackle this. In our experience, given the high accuracies observed on our trained models, we expect the conditional probabilities to be learned quite well, thus we expect a rather low count of samples that do not follow such implicit rules from the original data. In any case, we expect it to be way lower than other ML methods without codified rejection. 

Some of those rules are already tackled in our SDK (eg `a value multiple of 1000` can be achieved by using TABULAR_NUMERIC_DIGIT encoding, it will stick to sampling only zeros for the last 3 digits)

Do you have some examples where we could see a high rate of such occurrences ?

**Comment by spreeni** [comment]
@radurogojanumai Thanks for the fast reply! 

You are right, I did not consider `TABULAR_NUMERIC_DIGIT` for discrete increments.

I am seeing this currently as I am still forced to work on a smaller subset for now and would like to enforce some limits. If I duplicate that test set, the accuracy goes up but of course statistical slips still happen. So here I will need post-synthesis validation now. I think it could be a nice convenience functionality to provide this optionally, but definitely possible to work around in the current SDK.

And I imagine that this will be less of an issue once we scale the training data to larger sizes.
